heartbeat/gcp-vpc-move-ip

```diff
@@ -0,0 +1,302 @@
+      project=project, zone=zone, instance=instance)
+  return request.execute()
+
+
+def get_network_ifaces(project, zone, instance):
+  return get_instance(project, zone, instance)['networkInterfaces']
+
+
+def wait_for_operation(project, zone, operation):
+  while True:
+    result = CONN.zoneOperations().get(
+        project=project,
+        zone=zone,
+        operation=operation['name']).execute()
+
+    if result['status'] == 'DONE':
+      if 'error' in result:
+        raise Exception(result['error'])
+      return
+    time.sleep(1)
+
+
+def set_alias(project, zone, instance, alias, alias_range_name=None):
+  fingerprint = get_network_ifaces(project, zone, instance)[0]['fingerprint']
+  body = {
+      'aliasIpRanges': [],
+      'fingerprint': fingerprint
+  }
+  if alias:
+    obj = {'ipCidrRange': alias}
+    if alias_range_name:
+      obj['subnetworkRangeName'] = alias_range_name
+    body['aliasIpRanges'].append(obj)
+
+  request = CONN.instances().updateNetworkInterface(
+      instance=instance, networkInterface='nic0', project=project, zone=zone,
+      body=body)
+  operation = request.execute()
+  wait_for_operation(project, zone, operation)
+
+
+def get_alias(project, zone, instance):
+  iface = get_network_ifaces(project, zone, instance)
+  try:
+    return iface[0]['aliasIpRanges'][0]['ipCidrRange']
+  except KeyError:
+    return ''
+
+
+def get_localhost_alias():
+  net_iface = get_metadata('instance/network-interfaces', {'recursive': True})
+  net_iface = json.loads(net_iface.decode('utf-8'))
+  try:
+    return net_iface[0]['ipAliases'][0]
+  except (KeyError, IndexError):
+    return ''
+
+
+def get_zone(project, instance):
+  request = CONN.instances().aggregatedList(project=project)
+  while request is not None:
+    response = request.execute()
+    zones = response.get('items', {})
+    for zone in zones.values():
+      for inst in zone.get('instances', []):
+        if inst['name'] == instance:
+          return inst['zone'].split("/")[-1]
+    request = CONN.instances().aggregatedList_next(
+        previous_request=request, previous_response=response)
+  raise Exception("Unable to find instance %s" % (instance))
+
+
+def gcp_alias_start(alias):
+  if not alias:
+    sys.exit(OCF_ERR_CONFIGURED)
+  my_alias = get_localhost_alias()
+  my_zone = get_metadata('instance/zone').split('/')[-1]
+  project = get_metadata('project/project-id')
+
+  # If I already have the IP, exit. If it has an alias IP that isn't the VIP,
+  # then remove it
+  if my_alias == alias:
+    LOGGER.info(
+        '%s already has %s attached. No action required' % (HOSTNAME, alias))
+    sys.exit(OCF_SUCCESS)
+  elif my_alias:
+    LOGGER.info('Removing %s from %s' % (my_alias, HOSTNAME))
+    set_alias(project, my_zone, HOSTNAME, '')
+
+  # Loops through all hosts & remove the alias IP from the host that has it
+  hostlist = os.environ.get('OCF_RESKEY_hostlist', '')
+  hostlist.replace(HOSTNAME, '')
+  for host in hostlist.split():
```

Comment:
Isnt there a way to check if an instance is already using the IP?

Reply:
Yes, I am already doing that here: https://github.com/ClusterLabs/resource-agents/pull/1167/files#diff-5658f80e769d4fee411e8db09ee41c2dR191
Or do you mean something else?

Reply:
Yeah, I mean so we can avoid having to provide a hostlist, as that's inconvenient, so we usually avoid it when possible.

Reply:
I see. I could retrieve a list of all the machines in the GCE project, and query the IP in all of them if hostlist is not provided, what do you think?

Reply:
That'd be great. Hopefully we can get the information on which node is having the IP in the not to distant future as well.

Reply:
done